### PR TITLE
Swap in pmorie/go-open-service-broker-client

### LINKED ISSUE #940: Migrate to pmorie/go-open-service-broker-client
This issue is to migrate the controller code and unit tests to the https://github.com/pmorie/go-open-service-broker-client repository, as discussed in the 20170612 and 20170605 SIG meetings.

### PR DESCRIPTION
**Pull request by pmorie**
Closes #940

### CODE REVIEW

File: pkg/controller/controller_instance.go [hunk 1]
```diff
@@ -320,49 +326,59 @@ func (c *controller) pollInstance(serviceClass *v1alpha1.ServiceClass, servicePl
...
-	resp, rc, err := brokerClient.PollServiceInstance(instance.Spec.ExternalID, lastOperationRequest)
+
+	glog.V(5).Infof("Polling last operation on Instance %v/%v", instance.Namespace, instance.Name)
+
+	response, err := brokerClient.PollLastOperation(request)
 	if err != nil {
+		// If the operation was for delete and we receive a http.StatusGone,
+		// this is considered a success as per the spec, so mark as deleted
+		// and remove any finalizers.
+		if osbclient.IsGoneError(err) && deleting {
```
Comment (MHBauer) [inline pkg/controller/controller_instance.go]: @jwforres I think this fixes #953.

**Comment by MHBauer** [comment]
needs to be rebased before travis will pick it up.

**Comment by pmorie** [comment]
@MHBauer that is a terrible UX

File: pkg/controller/controller_test.go [hunk 2]
```diff
@@ -626,106 +596,144 @@ func checkPlan(serviceClass *v1alpha1.ServiceClass, index int, planName, planDes
...
-	}
-	if !foundSvcMeta2 {
-		t.Fatalf("Didn't find metadata in service2")
-	}
-	if !foundPlanMeta {
-		t.Fatalf("Didn't find metadata '' in service1 plan2")
-	}
+// FIX: there is an inconsistency between the current broker API types re: the
+// Service.Metadata field.  Our repo types it as `interface{}`, the go-open-
+// service-broker-client types it as `map[string]interface{}`.
```
Comment (arschles) [inline pkg/controller/controller_test.go]: can you create an issue to resolve this in a follow-up?

File: pkg/controller/controller_test.go [hunk 3]
```diff
@@ -976,66 +978,7 @@ func newTestController(t *testing.T) (
...
-	}
-
-	return &testControllerWithBrokerServer{
-		FakeKubeClient:      fakeKubeClient,
-		FakeCatalogClient:   fakeCatalogClient,
-		Controller:          testController.(*controller),
-		Informers:           serviceCatalogSharedInformers,
-		BrokerServerHandler: brokerHandler,
-		BrokerServer:        brokerServer,
-	}, nil
```
Comment (arschles) [inline pkg/controller/controller_test.go]: why remove this? I think you changed the single test that relied on it, but we have plans to use the fake broker server in the future, in integration tests and select unit tests (like in https://github.com/kubernetes-incubator/service-catalog/pull/923)

File: pkg/controller/controller_test.go [hunk 4]
```diff
@@ -1379,3 +1322,94 @@ func assertEmptyFinalizers(t *testing.T, obj runtime.Object) {
...
+	}
+}
+
+func assertProvision(t *testing.T, action fakeosb.Action, request *osb.ProvisionRequest) {
+	if e, a := fakeosb.ProvisionInstance, action.Type; e != a {
+		fatalf(t, "unexpected action type; expected %v, got %v", e, a)
+	}
+
+	if e, a := request, action.Request; !reflect.DeepEqual(e, a) {
+		fatalf(t, "unexpected diff in provision request: %v\nexpected %+v\ngot      %+v", diff.ObjectReflectDiff(e, a), e, a)
```
Comment (arschles) [inline pkg/controller/controller_test.go]: nit: can you remove all those spaces in front of the 2nd `%+v`? maybe replace them with a newline

File: pkg/controller/controller_broker_test.go [hunk 5]
```diff
@@ -18,23 +18,34 @@ package controller
...
+	osb "github.com/pmorie/go-open-service-broker-client/v2"
+	fakeosb "github.com/pmorie/go-open-service-broker-client/v2/fake"
+
 	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1alpha1"
-	"github.com/kubernetes-incubator/service-catalog/pkg/brokerapi"
-	fakebrokerapi "github.com/kubernetes-incubator/service-catalog/pkg/brokerapi/fake"
-	fakebrokerserver "github.com/kubernetes-incubator/service-catalog/pkg/brokerapi/fake/server"
+	// "github.com/kubernetes-incubator/service-catalog/pkg/brokerapi"
+	// fakebrokerapi "github.com/kubernetes-incubator/service-catalog/pkg/brokerapi/fake"
+	// fakebrokerserver "github.com/kubernetes-incubator/service-catalog/pkg/brokerapi/fake/server"
```
Comment (pmorie) [inline pkg/controller/controller_broker_test.go]: oops

File: pkg/controller/controller_broker_test.go [hunk 6]
```diff
@@ -18,23 +18,34 @@ package controller
...
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/diff"
 
 	"k8s.io/client-go/pkg/api"
 	"k8s.io/client-go/pkg/api/v1"
 	clientgotesting "k8s.io/client-go/testing"
 )
 
+func init() {
+	flag.Set("v", "3")
```
Comment (pmorie) [inline pkg/controller/controller_broker_test.go]: debugging cruft that needs to be removed

File: pkg/controller/controller_broker_test.go [hunk 7]
```diff
@@ -112,59 +123,18 @@ func TestShouldReconcileBroker(t *testing.T) {
...
-		t.Fatalf("Received unexpected event: %v", a)
-	}
-
-	if controllerParams.BrokerServerHandler.CatalogRequests != 1 {
-		t.Fatalf(
-			"expected 1 catalog request, got %d",
-			controllerParams.BrokerServerHandler.CatalogRequests,
-		)
-	}
-}
```
Comment (pmorie) [inline pkg/controller/controller_broker_test.go]: it's replaced, yeah

File: pkg/controller/controller_broker_test.go [hunk 8]
```diff
@@ -261,14 +241,16 @@ func TestReconcileBrokerDelete(t *testing.T) {
 	// 0. Deleting the associated ServiceClass
 	// 1. Updating the ready condition
 	// 2. Removing the finalizer
-	assertNumberOfActions(t, actions, 3)
+	assertNumberOfActions(t, actions, 4)
```
Comment (pmorie) [inline pkg/controller/controller_broker_test.go]: oops

File: pkg/controller/controller_broker_test.go [hunk 9]
```diff
@@ -299,22 +288,27 @@ func TestReconcileBrokerErrorFetchingCatalog(t *testing.T) {
 	events := getRecordedEvents(testController)
 	assertNumEvents(t, events, 1)
 
-	expectedEvent := api.EventTypeWarning + " " + errorFetchingCatalogReason + " " + "Error getting broker catalog for broker \"test-broker\": instance not found"
+	expectedEvent := api.EventTypeWarning + " " + errorFetchingCatalogReason + " " + "Error getting broker catalog for broker \"test-broker\": ooops"
```
Comment (pmorie) [inline pkg/controller/controller_broker_test.go]: it's the error that was returned

File: pkg/controller/controller_test.go (hunk 3 above)
Comment (pmorie) [inline pkg/controller/controller_test.go]: I'll add it back in if you want

File: pkg/controller/controller_test.go (hunk 4 above)
Comment (pmorie) [inline pkg/controller/controller_test.go]: Those spaces are so the two lines line up -- makes it much easier to read the output when the test fails.

**Comment by jwforres** [comment]
This does seem to fix https://github.com/kubernetes-incubator/service-catalog/issues/953

File: pkg/controller/controller_broker_test.go [hunk 10]
```diff
@@ -377,40 +374,41 @@ func TestReconcileBrokerWithAuthError(t *testing.T) {
...
-	if e, a := "get", getAction.GetVerb(); e != a {
-		t.Fatalf("Unexpected verb on action; expected %v, got %v", e, a)
-	}
+	assertNumberOfActions(t, kubeActions, 0)
 
 	events := getRecordedEvents(testController)
 	assertNumEvents(t, events, 1)
 
-	expectedEvent := api.EventTypeWarning + " " + errorAuthCredentialsReason + " " + "Error getting broker auth credentials for broker \"test-broker\": auth secret didn't contain username"
+	expectedEvent := api.EventTypeWarning + " " + errorSyncingCatalogReason + ` Error reconciling serviceClass "test-serviceclass" (broker "test-broker"): error creating serviceclass`
```
Comment (pmorie) [inline pkg/controller/controller_broker_test.go]: sure - this is basically how every other test does it, and i want to clean up every other test too, follow-up okay?

Comment (arschles) [inline pkg/controller/controller_broker_test.go]: yup, follow-up is ok

File: pkg/controller/controller_test.go (hunk 4 above)
Comment (arschles) [inline pkg/controller/controller_test.go]: ok, fine with me

**Comment by arschles** [comment]
@jwforres that's awesome. Any idea why it fixes?

Also, @pmorie can you put `Closes #953` in the OP so that issue is closed when this PR is merged?

File: pkg/controller/controller_binding_test.go [hunk 11]
```diff
@@ -153,60 +167,64 @@ func TestReconcileBindingWithParameters(t *testing.T) {
...
-	// There should only be one action that says binding was created
 	updatedBinding := assertUpdateStatus(t, actions[0], binding)
 	assertBindingReadyTrue(t, updatedBinding)
 
-	updateObject, ok := updatedBinding.(*v1alpha1.Binding)
+	kubeActions := fakeKubeClient.Actions()
+	assertNumberOfActions(t, kubeActions, 3)
+
+	// first action is a get on the namespace
+	// second action is a get on the secret
```
Comment (arschles) [inline pkg/controller/controller_binding_test.go]: recapping what @pmorie and I talked about offline, this will be done in a follow-up

File: pkg/controller/controller_binding_test.go [hunk 12]
```diff
@@ -274,24 +316,71 @@ func TestReconcileBindingNonbindableServiceClassBindablePlan(t *testing.T) {
...
-
-	// There should only be one action that says binding was created
 	updatedBinding := assertUpdateStatus(t, actions[0], binding)
 	assertBindingReadyTrue(t, updatedBinding)
 
+	kubeActions := fakeKubeClient.Actions()
+	assertNumberOfActions(t, kubeActions, 3)
+
+	// first action is a get on the namespace
+	// second action is a get on the secret
```
Comment (arschles) [inline pkg/controller/controller_binding_test.go]: recapping what @pmorie and I talked about offline, this will be done in a follow-up

**Comment by pmorie** [comment]
Hrm, I'm not sure that I think this should do anything to meaningfully change #953

File: pkg/controller/controller_broker_test.go (hunk 9 above)
Comment (arschles) [inline pkg/controller/controller_broker_test.go]: Also, I'm ok with doing the cleanup in a follow-up, in favor of getting this improvement in as a foundation

File: pkg/controller/controller_test.go (hunk 2 above)
Comment (pmorie) [inline pkg/controller/controller_test.go]: https://github.com/kubernetes-incubator/service-catalog/issues/959

**Comment by pmorie** [comment]
I think this PR is as ready as it's going to get.

File: pkg/controller/controller_instance.go [hunk 13]
```diff
@@ -370,81 +369,133 @@ func (c *controller) pollInstance(serviceClass *v1alpha1.ServiceClass, servicePl
...
-	glog.V(4).Infof("Poll for %v/%v returned %q : %q", instance.Namespace, instance.Name, resp.State, resp.Description)
+		// If the operation was for delete and we receive a http.StatusGone,
+		// this is considered a success as per the spec, so mark as deleted
+		// and remove any finalizers.
+		if osb.IsGoneError(err) && deleting {
+			clone, err := api.Scheme.DeepCopy(instance)
+			if err != nil {
+				return err
+			}
+			toUpdate := clone.(*v1alpha1.Instance)
```
Comment (arschles) [inline pkg/controller/controller_instance.go]: I will tackle this in a follow-up

File: pkg/controller/controller_instance.go [hunk 14]
```diff
@@ -370,81 +369,133 @@ func (c *controller) pollInstance(serviceClass *v1alpha1.ServiceClass, servicePl
...
-		// this gets updated as a side effect in both cases below.
-		instance.Status.AsyncOpInProgress = false
+	case osb.StateSucceeded:
+		// Update the instance to reflect that an async operation is no longer
+		// in progress.
+		clone, err := api.Scheme.DeepCopy(instance)
+		if err != nil {
+			return err
+		}
+		toUpdate := clone.(*v1alpha1.Instance)
```
Comment (arschles) [inline pkg/controller/controller_instance.go]: I will tackle this in a follow-up

File: pkg/controller/controller_instance.go [hunk 15]
```diff
@@ -370,81 +389,133 @@ func (c *controller) pollInstance(serviceClass *v1alpha1.ServiceClass, servicePl
...
+	response, err := brokerClient.PollLastOperation(request)
 	if err != nil {
-		glog.Warningf("Poll failed for %v/%v  : %s", instance.Namespace, instance.Name, err)
-		return err
-	}
-	glog.V(4).Infof("Poll for %v/%v returned %q : %q", instance.Namespace, instance.Name, resp.State, resp.Description)
+		// If the operation was for delete and we receive a http.StatusGone,
+		// this is considered a success as per the spec, so mark as deleted
+		// and remove any finalizers.
+		if osb.IsGoneError(err) && deleting {
```
Comment (MHBauer) [inline pkg/controller/controller_instance.go]: @jwforres @arschles @pmorie we didn't check for "is gone" before.

Comment (MHBauer) [inline pkg/controller/controller_instance.go]: This is what fixes #953

File: glide.lock [hunk 16]
```diff
@@ -200,427 +72,29 @@ imports:
...
-  - pkg/util/proxy
-  - pkg/util/trace
-  - pkg/util/trie
-  - pkg/util/webhook
-  - pkg/util/wsstream
-  - plugin/pkg/authenticator/token/webhook
-  - plugin/pkg/authorizer/webhook
 - name: k8s.io/client-go
   version: 76153773eaa3a268131d3d993290a194a1370585
-  subpackages:
```
Comment (MHBauer) [inline glide.lock]: one of us is using a different version of glide, or using it outside of the container. seeing as it's not normally run, I don't particularly care, but I find the churn here confusing.

File: pkg/controller/controller_instance.go (hunk 15 above)
Comment (pmorie) [inline pkg/controller/controller_instance.go]: @MHBauer that shouldn't affect anything for bindings, since they are only synchronous right now.

File: glide.lock (hunk 16 above)
Comment (pmorie) [inline glide.lock]: it is also confusing to me; let me upgrade glide and see where that gets me.

File: pkg/controller/controller.go [hunk 17]
```diff
@@ -349,39 +371,42 @@ func (c *controller) getServiceClassPlanAndBrokerForBinding(instance *v1alpha1.I
 // contained in the secret referenced in the Broker's AuthSecret field, or
 // returns an error. If the AuthSecret field is nil, empty values are
```
Comment (MHBauer) [inline pkg/controller/controller.go]: non-blocking issue: comment needs to be updated in the past.

File: pkg/controller/controller_binding.go [hunk 18]
```diff
@@ -205,14 +206,18 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
...
+			InstanceID:   instance.Spec.ExternalID,
 			ServiceID:    serviceClass.ExternalID,
 			PlanID:       servicePlan.ExternalID,
+			AppGUID:      &appGUID,
 			Parameters:   parameters,
-			AppGUID:      string(ns.UID),
-			BindResource: map[string]interface{}{"app_guid": string(ns.UID)},
+			BindResource: &osb.BindResource{AppGUID: &appGUID},
 		}
-		response, err := brokerClient.CreateServiceBinding(instance.Spec.ExternalID, binding.Spec.ExternalID, request)
```
Comment (MHBauer) [inline pkg/controller/controller_binding.go]: much cleaner. this is some frankenstein stuff right here.

File: pkg/controller/controller.go (hunk 17 above)
Comment (MHBauer) [inline pkg/controller/controller.go]: It's called AuthInfo, with a BasicAuthSecret internal field.

Comment (pmorie) [inline pkg/controller/controller.go]: ahhhhhh

File: pkg/controller/controller_broker_test.go [hunk 19]
```diff
@@ -22,13 +22,16 @@ import (
 	"testing"
 	"time"
 
+	_ "github.com/golang/glog"
```
Comment (MHBauer) [inline pkg/controller/controller_broker_test.go]: why the empty import?

Comment (pmorie) [inline pkg/controller/controller_broker_test.go]: Good catch, it's cruft I missed

File: test/integration/controller_test.go [hunk 20]
```diff
@@ -227,34 +255,41 @@ func TestBasicFlows(t *testing.T) {
...
+//
+// - a fake kubernetes core api client
+// - a fake service catalog api client
+// - a fake osb client
+// - a test controller
+// - the shared informers for the service catalog v1alpha1 api
+//
+// If there is an error, newTestController calls 'Fatal' on the injected
+// testing.T.
+func newTestController(t *testing.T, config fakeosb.FakeClientConfiguration) (
```
Comment (MHBauer) [inline test/integration/controller_test.go]: nice docs. nice dependency injection.

File: pkg/controller/controller_broker_test.go [hunk 21]
```diff
@@ -260,15 +230,18 @@ func TestReconcileBrokerDelete(t *testing.T) {
 	// The three actions should be:
 	// 0. Deleting the associated ServiceClass
 	// 1. Updating the ready condition
-	// 2. Removing the finalizer
-	assertNumberOfActions(t, actions, 3)
+	// 2. Getting the broker
+	// 3. Removing the finalizer
+	assertNumberOfActions(t, actions, 4)
```
Comment (MHBauer) [inline pkg/controller/controller_broker_test.go]: I still like this comment! 👍

File: pkg/controller/controller_instance.go (hunk 15 above)
Comment (MHBauer) [inline pkg/controller/controller_instance.go]: I thought I already responded to this, but

I know that the equivalent code was being run on #953. I know that we weren't checking for IsGone, which is what was being returned 410. So, we do need to add a test for this, and I added something like it in #955, but yes, more testing if we can get it to fail again.

File: pkg/controller/controller_instance.go [hunk 22]
```diff
@@ -370,81 +389,133 @@ func (c *controller) pollInstance(serviceClass *v1alpha1.ServiceClass, servicePl
...
+
+		// We got some kind of error from the broker.  While polling last
+		// operation, this represents an invalid response and we should
+		// continue polling last operation.
+		//
+		// The ready condition on the instance should already have
+		// condition false; it should be sufficient to create an event for
+		// the instance.
+		errText := ""
+		if osb.IsHTTPError(err) {
```
Comment (MHBauer) [inline pkg/controller/controller_instance.go]: general comment is that we're creating a giant state machine here. We need to write it up and draw a picture and refactor this code to be smaller more digestible pieces.

File: pkg/controller/controller_instance_test.go [hunk 23]
```diff
@@ -935,17 +1017,56 @@ func TestPollServiceInstanceStatusGoneDeprovisioningWithOperationNoFinalizer(t *
 	assertNumEvents(t, events, 1)
 }
 
-func TestPollServiceInstanceSuccessDeprovisioningWithOperationWithFinalizer(t *testing.T) {
-	fakeKubeClient, fakeCatalogClient, fakeBrokerClient, testController, sharedInformers := newTestController(t)
-
-	fakeBrokerClient.CatalogClient.RetCatalog = getTestCatalog()
+func TestPollServiceInstanceBrokerError(t *testing.T) {
```
Comment (MHBauer) [inline pkg/controller/controller_instance_test.go]: new test or diff freaking out on me?

File: pkg/controller/controller_test.go (hunk 2 above)
Comment (MHBauer) [inline pkg/controller/controller_test.go]: eww, but good to have a followup.

File: pkg/controller/controller_test.go [hunk 24]
```diff
@@ -769,8 +779,12 @@ func falsePtr() *bool {
 	return &b
 }
 
+func strPtr(s string) *string {
```
Comment (MHBauer) [inline pkg/controller/controller_test.go]: still think this is weird.

File: pkg/controller/controller_instance_test.go (hunk 23 above)
Comment (pmorie) [inline pkg/controller/controller_instance_test.go]: new test

File: pkg/controller/controller_instance.go (hunk 22 above)
Comment (pmorie) [inline pkg/controller/controller_instance.go]: Yeah, we are definitely approaching the point where it will become a necessity to apply extract method and break stuff up.  I was pretty militant about _not_ doing that when a lot of people in the group were learning the controller pattern.  The reason for that is that extract-method application tends to solidify a design in a lot of people's heads and shape how they think about a problem space.  I've seen that go very badly in other parts of kubernetes (and in other projects in my career), and the way we remedied it was by refactoring to this exploded style of factoring which is quite long-winded, but fairly clear in the sense of what the code is doing.

Basically, I wanted to avoid some mistake in applying extract-method result in the mentality of 'we can't do it X way, we'd have to change all this code in the controller'!  Having code that literally no one is happy with helps people developing mental barricades against a design that might be better because they've fallen in love with a refactor PR they did.

If we can possibly, possibly hang on, I'd prefer to leave it this way (the mental handle I have for this style of code is 'unfactored') until we have the OSB integration completed (ie, plan update, param update).  I'm not sure if that will be possible, though.

LMK if this comment makes any sense.

Comment (MHBauer) [inline pkg/controller/controller_instance.go]: Some sense made. It's less extract method and more be really explicit about state transitions. If I come in here, I basically need to look for "state is my current state" before I can start parsing. but i've got to mentally go through and discard 85% of the code by reading if this if that. not even if-else.

I'd almost prefer if it was if-else and resubmit to queue, because then I could find the "one clause" that I need that could possibly execute.

I'm not advocating single exit policy, but it's very complicated here, and we're really not doing anything very complex.
